File: scripts/figures/plot_sphere_3d_convergence.jl
#!/usr/bin/env julia --project=docs
# WP-3D-6 — Figure: sphere 3D Re=20 convergence (uniform vs stretched).
# Julia/CairoMakie version of plot_sphere_3d_convergence.py.
#
# Usage:
#   julia --project=docs scripts/figures/plot_sphere_3d_convergence.jl \
#       --log results/slbm_sphere_3d.log \
#       --out paper/figures/sphere_3d_convergence.pdf

using CairoMakie

const LINE_RE = r"^(?<label>\S.*?)\s+(?<Nx>\d+)×(?<Ny>\d+)×(?<Nz>\d+)\s+\(\s*(?<cells>\d+)\s*cells\)\s+Cd=(?<cd>[-\d.]+)\s+err=(?<err>[-\d.]+)%\s+MLUPS=(?<mlups>[-\d.]+)"

struct Run
    label::String
    cells::Int
    Cd::Float64
    err::Float64
    mlups::Float64
end

function parse_log(path::AbstractString)
    rows = Run[]
    for raw in eachline(path)
        m = match(LINE_RE, strip(raw))
        m === nothing && continue
        push!(rows, Run(strip(m["label"]),
                         parse(Int, m["cells"]),
                         parse(Float64, m["cd"]),
                         parse(Float64, m["err"]),
                         parse(Float64, m["mlups"])))
    end
    return rows
end

function split_runs(rows)
    uni  = filter(r -> startswith(lowercase(r.label), "uniform"),  rows)
    strd = filter(r -> startswith(lowercase(r.label), "stretch"), rows)
    sort!(uni;  by = r -> r.cells)
    sort!(strd; by = r -> r.cells)
    return uni, strd
end

function build_fig(uni::Vector{Run}, strd::Vector{Run}, out_path::AbstractString)
    fig = Figure(; size = (900, 360))

    ax_err = Axis(fig[1, 1];
        xlabel = "cells",
        ylabel = "|Cd − Cd_ref| / Cd_ref [%]",
        title  = "Sphere Re=20 — Cd accuracy",
        xscale = log10, yscale = log10,
        xgridstyle = :dash, ygridstyle = :dash)
    if !isempty(uni)
        scatterlines!(ax_err, [r.cells for r in uni], [r.err for r in uni];
                      color = :steelblue, marker = :circle,
                      markersize = 12, linewidth = 2,
                      label = "Uniform Cartesian")
    end
    if !isempty(strd)
        scatterlines!(ax_err, [r.cells for r in strd], [r.err for r in strd];
                      color = :darkorange, marker = :rect,
                      markersize = 12, linewidth = 2, linestyle = :dash,
                      label = "Stretched (preliminary)")
    end
    axislegend(ax_err; position = :rt)

    ax_mlups = Axis(fig[1, 2];
        xlabel = "cells",
        ylabel = "MLUPS (H100, Float64)",
        title  = "Throughput vs grid size",
        xscale = log10,
        xgridstyle = :dash, ygridstyle = :dash)
    if !isempty(uni)
        scatterlines!(ax_mlups, [r.cells for r in uni], [r.mlups for r in uni];
                      color = :steelblue, marker = :circle,
                      markersize = 12, linewidth = 2,
                      label = "Uniform Cartesian")
    end
    if !isempty(strd)
        scatterlines!(ax_mlups, [r.cells for r in strd], [r.mlups for r in strd];
                      color = :darkorange, marker = :rect,
                      markersize = 12, linewidth = 2, linestyle = :dash,
                      label = "Stretched (preliminary)")
    end
    axislegend(ax_mlups; position = :rb)

    mkpath(dirname(out_path))
    save(out_path, fig)
    @info "wrote" out_path
end

function main(args)
    log_path = "paper/data/sphere_3d_h100_20147745.log"
    out_path = "paper/figures/sphere_3d_convergence.pdf"
    i = 1
    while i ≤ length(args)
        if args[i] == "--log" && i + 1 ≤ length(args)
            log_path = args[i + 1]; i += 2
        elseif args[i] == "--out" && i + 1 ≤ length(args)
            out_path = args[i + 1]; i += 2
        else
            i += 1
        end
    end
    isfile(log_path) || error("log file not found: $log_path")
    rows = parse_log(log_path)
    isempty(rows) && error("no benchmark lines parsed; check log format")
    uni, strd = split_runs(rows)
    println("parsed $(length(rows)) runs ($(length(uni)) uniform, $(length(strd)) stretched)")
    for r in rows
        println("  ", rpad(r.label, 30), "  cells=", lpad(r.cells, 9),
                "  Cd=", round(r.Cd, digits = 3),
                "  err=", round(r.err, digits = 2), "%",
                "  MLUPS=", round(Int, r.mlups))
    end
    build_fig(uni, strd, out_path)
end

main(ARGS)

File: scripts/figures/plot_sphere_3d_convergence.py
#!/usr/bin/env python3
"""
WP-3D-6 — Figure: sphere 3D Re=20 convergence (uniform vs stretched).

Parses the structured output of `hpc/slbm_sphere_h100.jl` and produces:
  paper/figures/sphere_3d_convergence.pdf — Cd error and MLUPS vs cells

Expected log line format (one per run):
    <label> <Nx>x<Ny>x<Nz> ( <cells> cells) Cd=<val> err=<val>% MLUPS=<val> (<elapsed>s, ...)

Usage:
    python scripts/figures/plot_sphere_3d_convergence.py \
        --log results/slbm_sphere_3d.log \
        --out paper/figures/sphere_3d_convergence.pdf
"""

import argparse
import re
import sys
from pathlib import Path

import matplotlib.pyplot as plt

LINE_RE = re.compile(
    r"^(?P<label>\S.*?)\s+"
    r"(?P<Nx>\d+)x(?P<Ny>\d+)x(?P<Nz>\d+)\s+"
    r"\(\s*(?P<cells>\d+)\s*cells\)\s+"
    r"Cd=(?P<cd>[-\d.]+)\s+"
    r"err=(?P<err>[-\d.]+)%\s+"
    r"MLUPS=(?P<mlups>[-\d.]+)"
)


def parse(path: Path):
    rows = []
    for raw in path.read_text().splitlines():
        m = LINE_RE.search(raw.strip())
        if not m:
            continue
        rows.append({
            "label":  m["label"].strip(),
            "Nx":     int(m["Nx"]),
            "Ny":     int(m["Ny"]),
            "Nz":     int(m["Nz"]),
            "cells":  int(m["cells"]),
            "Cd":     float(m["cd"]),
            "err":    float(m["err"]),
            "mlups":  float(m["mlups"]),
        })
    return rows


def split_uniform_stretched(rows):
    uni  = [r for r in rows if r["label"].lower().startswith("uniform")]
    strd = [r for r in rows if r["label"].lower().startswith("stretch")]
    uni.sort(key=lambda r: r["cells"])
    strd.sort(key=lambda r: r["cells"])
    return uni, strd


def plot(uni, strd, out_path: Path):
    fig, (ax_err, ax_mlups) = plt.subplots(1, 2, figsize=(10.5, 4.0))

    if uni:
        ax_err.plot([r["cells"] for r in uni], [r["err"] for r in uni],
                    "o-", label="Uniform Cartesian", linewidth=2, markersize=7)
    if strd:
        ax_err.plot([r["cells"] for r in strd], [r["err"] for r in strd],
                    "s--", label="Stretched (local-CFL)", linewidth=2, markersize=7)
    ax_err.set_xscale("log")
    ax_err.set_yscale("log")
    ax_err.set_xlabel("cells")
    ax_err.set_ylabel(r"$|\mathrm{Cd} - \mathrm{Cd}_\mathrm{ref}| / \mathrm{Cd}_\mathrm{ref}$ [%]")
    ax_err.grid(True, which="both", alpha=0.3)
    ax_err.legend()
    ax_err.set_title("Sphere Re=20 — Cd accuracy")

    if uni:
        ax_mlups.plot([r["cells"] for r in uni], [r["mlups"] for r in uni],
                      "o-", label="Uniform Cartesian", linewidth=2, markersize=7)
    if strd:
        ax_mlups.plot([r["cells"] for r in strd], [r["mlups"] for r in strd],
                      "s--", label="Stretched (local-CFL)", linewidth=2, markersize=7)
    ax_mlups.set_xscale("log")
    ax_mlups.set_xlabel("cells")
    ax_mlups.set_ylabel("MLUPS (H100, Float64)")
    ax_mlups.grid(True, which="both", alpha=0.3)
    ax_mlups.legend()
    ax_mlups.set_title("Throughput vs grid size")

    fig.tight_layout()
    out_path.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(out_path, bbox_inches="tight")
    print(f"wrote {out_path}")


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--log", type=Path, required=True,
                   help="path to slbm_sphere_3d.log produced on Aqua")
    p.add_argument("--out", type=Path,
                   default=Path("paper/figures/sphere_3d_convergence.pdf"))
    args = p.parse_args()

    if not args.log.exists():
        sys.exit(f"log file not found: {args.log}")

    rows = parse(args.log)
    if not rows:
        sys.exit("no benchmark lines parsed; check log format")

    uni, strd = split_uniform_stretched(rows)
    print(f"parsed {len(rows)} runs ({len(uni)} uniform, {len(strd)} stretched)")
    for r in rows:
        print(f"  {r['label']:30s}  cells={r['cells']:>10d}  "
              f"Cd={r['Cd']:.3f}  err={r['err']:.2f}%  MLUPS={r['mlups']:.0f}")

    plot(uni, strd, args.out)


if __name__ == "__main__":
    main()
